### PR TITLE
feature/#695 added check for pipeline task exist in scenario

### PR DESCRIPTION
**Pull request by toan-quach**
#695 

This PR is set to Draft only and please don't merge as it will merge directly to branch `feature/#718-stop-storing-PipelineModel` instead of `develop`. I will create a new PR later to merge directly to develop after PR [#722](https://github.com/Avaiga/taipy-core/pull/722) has been merged

### CODE REVIEW

File: src/taipy/core/exceptions/exceptions.py [hunk 1]
```diff
@@ -147,10 +150,17 @@ def __init__(self, pipeline_id: str, scenario_id: str):
 
 
 class PipelineTaskDoesNotExistInSameScenario(Exception):
-    """Raised if the Task of a Pipeline does not exist in the same Scenario."""
+    """Raised if the Tasks of a Pipeline does not exist in the same Scenario."""
```
Comment (trgiangdo) [inline src/taipy/core/exceptions/exceptions.py]:
```suggestion
    """Raised if Tasks of a Pipeline do not exist in the same Scenario that the Pipeline belongs to."""
```
Or
```suggestion
    """Raised if there is any Task of a Pipeline does not exist in the same Scenario that the Pipeline belongs to."""
```

File: src/taipy/core/exceptions/exceptions.py [hunk 2]
```diff
@@ -147,10 +150,17 @@ def __init__(self, pipeline_id: str, scenario_id: str):
...
+
+    def __init__(self, task_ids: List[Union[str, None]], pipeline_name: str, scenario_id: str):
```
Comment (trgiangdo) [inline src/taipy/core/exceptions/exceptions.py]:
```suggestion
    def __init__(self, task_ids: List[str], pipeline_name: str, scenario_id: str):
```
The type hint here is weird, it should not be a List of None. Is this what you meant?

File: src/taipy/core/exceptions/exceptions.py [hunk 3]
```diff
@@ -147,10 +150,17 @@ def __init__(self, pipeline_id: str, scenario_id: str):
...
+        self.message = f"Tasks {task_ids} of Pipeline {pipeline_name} does not exist in Scenario {scenario_id}."
+
+
+class PipelineTaskConfigDoesNotExistInSameScenarioConfig(Exception):
+    """Raised if the TaskConfigs of a Pipeline does not exist in the same ScenarioConfig."""
```
Comment (trgiangdo) [inline src/taipy/core/exceptions/exceptions.py]:
```suggestion
    """Raised if TaskConfigs of a Pipeline do not exist in the same ScenarioConfig that the Pipeline belongs to."""
```
Or 
```suggestion
    """Raised if there is any TaskConfig of a Pipeline does not exist in the same ScenarioConfig that the Pipeline belongs to."""
```

File: src/taipy/core/exceptions/exceptions.py [hunk 4]
```diff
@@ -147,10 +150,17 @@ def __init__(self, pipeline_id: str, scenario_id: str):
...
 
-    def __init__(self, task_id: str, pipeline_id: str, scenario_id: str):
-        self.message = f"Task {task_id} of Pipeline {pipeline_id} does not exist in Scenario {scenario_id}."
+    def __init__(self, task_config_ids: List[Union[str, None]], pipeline_name: str, scenario_config_id: str):
```
Comment (trgiangdo) [inline src/taipy/core/exceptions/exceptions.py]:
```suggestion
    def __init__(self, task_config_ids: List[str], pipeline_name: str, scenario_config_id: str):
```

File: tests/core/scenario/test_scenario.py [hunk 5]
```diff
@@ -226,6 +225,51 @@ def get_label(self):
...
+
+    scenario.tasks = [task_1]
+
+    scenario.add_pipelines({"pipeline_2": {"tasks": [task_1]}})
+
+    with pytest.raises(PipelineTaskDoesNotExistInSameScenario):
+        scenario.add_pipelines({"pipeline_3": {"tasks": [task_2]}})
+
+    with pytest.raises(PipelineTaskDoesNotExistInSameScenario):
+        scenario.add_pipelines({"pipeline_4": {"tasks": [task_1, task_2]}})
```
Comment (trgiangdo) [inline tests/core/scenario/test_scenario.py]: I think we should also check for the message of the Exception.

```suggestion
    with pytest.raises(PipelineTaskDoesNotExistInSameScenario) as error:
        scenario.add_pipelines({"pipeline_3": {"tasks": [task_2]}})
        assert "[task_2] ..." in error.message

    with pytest.raises(PipelineTaskDoesNotExistInSameScenario) as error:
        scenario.add_pipelines({"pipeline_4": {"tasks": [task_1, task_2]}})
        assert "[task_2] ..." in error.message
        assert "task_1" not in error.message
```

File: tests/core/scenario/test_scenario_manager.py [hunk 6]
```diff
@@ -220,6 +220,32 @@ def test_set_and_get_scenario(cycle):
...
+
+    scenario_config_1._tasks = [task_config_1]
+    _ScenarioManager._create(scenario_config_1)
+
+    scenario_config_1.add_sequences({"pipeline_2": [task_config_1]})
+    _ScenarioManager._create(scenario_config_1)
+
+    scenario_config_1.add_sequences({"pipeline_3": [task_config_1, task_config_2]})
+    with pytest.raises(PipelineTaskConfigDoesNotExistInSameScenarioConfig):
+        _ScenarioManager._create(scenario_config_1)
```
Comment (trgiangdo) [inline tests/core/scenario/test_scenario_manager.py]: Same here, I think we should also check for the message of the exception

**Review by trgiangdo — COMMENTED** [review]
I have a few comments

File: src/taipy/core/exceptions/exceptions.py (hunk 2 above)
Comment (toan-quach) [reply src/taipy/core/exceptions/exceptions.py]: ah you're right, I think it somehow got messed up with some of the automation tool for type hint

Comment (toan-quach) [reply src/taipy/core/exceptions/exceptions.py]: I can't seem to get mypy to agree with me after several if-else guaranteeing the list will not be None, so I settled with `task_ids: List[Optional[str]]` and `task_config_ids: List[Optional[str]]`

**Review by jrobinAV — DISMISSED** [review]

**Review by trgiangdo — APPROVED** [review]

**Pull request merged**